[AutoDiff] Reabstraction thunk differentiation.

dan-zheng opened the pull request:
Support reabstraction thunk differentiation.
Specifically, support this reabstraction thunk `partial_apply` pattern:

```
%generic_fn = function_ref @generic : $<τ_0_0> (@in_guaranteed τ_0_0) -> @out τ_0_0
%specialized_fn = partial_apply %generic_fn<Float>()
%reabs_thunk = function_ref @reabs_thunk : $(Float, @guaranteed @callee_guaranteed (@in_guaranteed Float) -> @out Float) -> Float
%thunked_fn = partial_apply %reabs_thunk(%specialized_fn)
%diff_fn = differentiable_function %thunked_fn
```

This enables differentiation of direct references to generic functions, which
generate reabstraction thunk `partial_apply`s in SIL.

- SIL: add special reabstraction thunk case to
  `SILFunctionType::getAutoDiffDerivativeFunctionType`.
  - Reabstraction thunks have a function-typed last argument, representing the
    function to reabstract.
  - That argument is transformed into a `@differentiable` function-typed
    argument for reabstraction thunk JVPs/VJPs. This enables differentiation
    transform support: reabstraction thunk JVP/VJP callers are responsible for
    passing a `@differentiable` function. Otherwise, the function-typed
    argument is opaque and cannot be differentiated.
- Differentiation transform:
  - Handle reabstraction thunk `partial_apply` reapplications in
    `reapplyFunctionConversion`. Form a `differentiable_function` of the
    function-typed argument and `partial_apply` it to the reabstraction thunk
    JVP/VJP.

Resolves TF-201. Short write-up about the approach [here](https://docs.google.com/document/d/1_BirmTqdotglwNTOcYAW-ib6mx_jl-gH9Dbg4WmHZh0/edit#bookmark=id.inyyvy1wapx).
Exposes TF-1033: differential generation ownership error.
TF-1036 tracks future reabstraction thunk derivative generation optimizations.

---

Example:

```swift
func id<T>(_ x: T) -> T { x }
print(gradient(at: Float(10), in: id))
// 1.0

protocol DifferentiableAdditiveArithmetic: Differentiable & AdditiveArithmetic {
  @differentiable
  static func + (lhs: Self, rhs: Self) -> Self
}
extension Float: DifferentiableAdditiveArithmetic {}
func generic<T: DifferentiableAdditiveArithmetic>(_ x: T) -> T {
  x + x + x
}
print(gradient(at: Float(10), in: generic))
// 3.0
```

Before:
```
<unknown>:0: error: expression is not differentiable
<unknown>:0: note: opaque non-'@differentiable' function is not differentiable
```

After: the program runs and prints as expected.

dan-zheng committed:
[AutoDiff] Differentiate partially-applied reabstraction thunks.

Support differentiation of partially-applied reabstraction thunks:

```
%generic_fn = function_ref @generic : $<τ_0_0> (@in_guaranteed τ_0_0) -> @out τ_0_0
%specialized_fn = partial_apply %generic_fn<Float>()
%reabs_thunk = function_ref @reabs_thunk : $(Float, @guaranteed @callee_guaranteed (@in_guaranteed Float) -> @out Float) -> Float
%thunked_fn = partial_apply %reabs_thunk(%specialized_fn)
%diff_fn = differentiable_function %thunked_fn
```

The enables differentiation of direct references to generic functions.

- SIL: add special reabstraction thunk case to
  `SILFunctionType::getAutoDiffDerivativeFunctionType`.
  - Reabstraction thunks have a function as the last argument.
  - Reabstraction thunk JVPs/VJPs have a `@differentiable` function as the last
    argument.
- Differentiation transform: handle partially-applied reabstraction thunks in
  `reapplyFunctionConversion`.

Resolves TF-201.

dan-zheng commented:
@rxwei: is this special handling for reabstraction thunks obsolete after the Differentiable Curry (https://github.com/apple/swift/pull/28516) is fully implemented?

rxwei commented:
Yes, this will be obsolete, but I think the benefits of this change outweighs waiting for differentiable curry. Could you please mark your changes clearly with `FIXME(SR-11882)` so that things will be obsoleted cleanly when SR-11882 is done?

In short, it would make me really really happy if it allows the following:
```swift
let f: @differentialbe (Float, Float) -> Float = (+)
```

dan-zheng commented:
> Yes, this will be obsolete, but I think the benefits of this change outweighs waiting for differentiable curry. Could you please mark your changes clearly with `FIXME([SR-11882](https://bugs.swift.org/browse/SR-11882))` so that things will be obsoleted cleanly when [SR-11882](https://bugs.swift.org/browse/SR-11882) is done?

Okay, will do!

> In short, it would make me really really happy if it allows the following:
> 
> ```swift
> let f: @differentialbe (Float, Float) -> Float = (+)
> ```

This code actually involves curry thunks, not reabstraction thunks, and appears to have been working for a long time. Here's the `-emit-silgen` output:
```
// f
sil_global hidden [let] @$s4plus1fyS2f_SftXGvp : $@differentiable @callee_guaranteed (Float, Float) -> Float

// main
sil [ossa] @main : $@convention(c) (Int32, UnsafeMutablePointer<Optional<UnsafeMutablePointer<Int8>>>) -> Int32 {
bb0(%0 : $Int32, %1 : $UnsafeMutablePointer<Optional<UnsafeMutablePointer<Int8>>>):
  alloc_global @$s4plus1fyS2f_SftXGvp             // id: %2
  %3 = global_addr @$s4plus1fyS2f_SftXGvp : $*@differentiable @callee_guaranteed (Float, Float) -> Float // user: %8
  %4 = metatype $@thin Float.Type                 // user: %6
  // function_ref curry thunk of static Float.+ infix(_:_:)
  %5 = function_ref @$sSf1poiyS2f_SftFZTc : $@convention(thin) (@thin Float.Type) -> @owned @callee_guaranteed (Float, Float) -> Float // user: %6
  %6 = apply %5(%4) : $@convention(thin) (@thin Float.Type) -> @owned @callee_guaranteed (Float, Float) -> Float // user: %7
  %7 = differentiable_function [parameters 0 1] %6 : $@callee_guaranteed (Float, Float) -> Float // user: %8
  store %7 to [init] %3 : $*@differentiable @callee_guaranteed (Float, Float) -> Float // id: %8
  %9 = integer_literal $Builtin.Int32, 0          // user: %10
  %10 = struct $Int32 (%9 : $Builtin.Int32)       // user: %11
  return %10 : $Int32                             // id: %11
} // end sil function 'main'
```

rxwei commented:
Awesome.

dan-zheng committed:
Refactor tests to avoid orthogonal issue TF-1033.

dan-zheng committed:
Merge branch 'tensorflow' of github.com:apple/swift into TF-201

dan-zheng committed:
Minor renaming for clarity.

dan-zheng committed:
Address review feedback.

- Add doc comment for `SILFunctionType::getAutoDiffDerivativeFunctionType` and
  explain reabstraction thunk special case.
- Change `reapplyFunctionConversion` to always take parameter indices.
  - Reabstraction thunk special logic is triggered when the original callee
    is a reabstraction thunk `function_ref`.

marcrasi commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -1248,20 +1247,30 @@ static void copyParameterArgumentsForApply(
     SmallVectorImpl<AllocStackInst *> &newBuffersToDealloc,
-    GenericSignature newFuncGenSig = GenericSignature()) {
+    GenericSignature newFuncGenSig = GenericSignature(),
+    IndexSubset *reabstractionThunkIndices = nullptr) {
```
Is there a reason why `reabstractionThunkIndices` is only set when dealing with reabstraction thunks? Would it make sense to make it a more general thing that we require callers to set in all cases?

I can't think of any reason why we'd need the information in other cases, but it would simplify the API and make mistakes where callers don't set the indices less likely.

dan-zheng replied:
Thanks for the suggestion!

https://github.com/apple/swift/pull/28570/commits/6eb7a070fd6734c1deb1220556038e4267d349ef changes `reapplyFunctionConversion` so that parameter indices are always passed. Reabstraction thunk special logic is triggered when the original callee is a reabstraction thunk `function_ref` - this is a nicer condition.

marcrasi commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -1248,20 +1247,30 @@ static void copyParameterArgumentsForApply(
+  if (auto *cvi = dyn_cast<CopyValueInst>(oldConvertedFunc)) {
+    auto innerNewFunc = reapplyFunctionConversion(
+        context, newFunc, oldFunc, cvi->getOperand(), builder, loc,
+        newBuffersToDealloc, newFuncGenSig);
```
do the `reabstractionThunkIndices` need to be passed into the recursive call here?

similar comment about all the recursive calls.

dan-zheng replied:
This hasn't been necessary in practice but seems good for robustness, e.g. to handle function conversion of the result of reabstraction thunk `partial_apply`s. I'll update this soon.

dan-zheng replied:
Done in https://github.com/apple/swift/pull/28570/commits/6eb7a070fd6734c1deb1220556038e4267d349ef.

dan-zheng commented on lib/SIL/SILFunctionType.cpp:
```diff
@@ -341,6 +341,20 @@ CanSILFunctionType SILFunctionType::getAutoDiffDerivativeFunctionType(
+  // Reabstraction thunks have a function-typed argument (the function to
+  // reabstract) as their last argument. Reabstraction thunk JVPs/VJPs have a
+  // `@differentiable` function-typed last argument instead.
+  if (isReabstractionThunk) {
```
Done in https://github.com/apple/swift/pull/28570/commits/6eb7a070fd6734c1deb1220556038e4267d349ef.

dan-zheng commented on include/swift/AST/Types.h:
```diff
@@ -4308,7 +4308,8 @@ class SILFunctionType final : public TypeBase, public llvm::FoldingSetNode,
       LookupConformanceFn lookupConformance,
-      CanGenericSignature derivativeFunctionGenericSignature = nullptr);
+      CanGenericSignature derivativeFunctionGenericSignature = nullptr,
+      bool isReabstractionThunk = false);
```
> It would be helpful to have a documentation comment explaining what the derivative function of a reabstraction thunk is.

Added doc comment for `SILFunctionType::getAutoDiffDerivativeFunctionType` in https://github.com/apple/swift/pull/28570/commits/6eb7a070fd6734c1deb1220556038e4267d349ef:
```cpp
  /// Special cases:
  /// - Reabstraction thunks have special derivative type calculation. The
  ///   original function-typed last parameter is transformed into a
  ///   `@differentiable` function-typed parameter in the derivative type. This
  ///   is necessary for the differentiation transform to support reabstraction
  ///   thunk differentiation because the function argument is opaque and cannot
  ///   be differentiated. Instead, the argument is made `@differentiable` and
  ///   reabstraction thunk JVP/VJP callers are reponsible for passing a
  ///   `@differentiable` function.
```

Also added comment in `reapplyFunctionConversion` in Differentiation.cpp explaining `partial_apply` reapplication for reabstraction thunk derivatives.

> If I understand correctly, it is "a function that, when partially applied to the original function, becomes the derivative function of the reabstracted function." Or maybe instead it "becomes a reabstraction of the derivative function". I haven't thought carefully enough to figure out which one it is, or if there is a difference, but maybe there are situations where the difference is important.

I believe your understanding is correct. I made a [short write-up](https://docs.google.com/document/d/1_BirmTqdotglwNTOcYAW-ib6mx_jl-gH9Dbg4WmHZh0/edit#bookmark=id.inyyvy1wapx) about the reabstraction thunk differentiation approach.

> Another question I have is: Why does the derivative function of the reabstraction thunk take the original function as argument rather than than the derivative function?

That's a good point! A reabstraction thunk derivative only needs the original function's corresponding derivative, not any other `@differentiable` function components. I think that approach is more efficient and semantically clear, so I'll pursue it now.

dan-zheng replied:
Actually, to focus on retrodiff as a priority, how about deferring optimizations and using the current approach for now? TF-1036 tracks reabstraction thunk derivative generation optimizations.

marcrasi replied:
sounds good

nice doc comment!

dan-zheng committed:
Reference TF-1036 optimizations.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng committed:
Merge branch 'tensorflow' of github.com:apple/swift into TF-201

dan-zheng committed:
Fix test based on upstream changes.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow
